File: ios/BiblosSurferTests/Services/BookStoreTests.swift
//
//  BookStoreTests.swift
//  BiblosSurferTests
//

import XCTest
@testable import BiblosSurfer

final class BookStoreTests: XCTestCase {
    func testProgressRoundTripsThroughTheStore() {
        let store = LocalBookService()
        let fileURL = URL(fileURLWithPath: "/tmp/Books/voyage.epub")
        let item = LibraryItem(fileURL: fileURL)

        store.upsert(item, relativePath: "voyage.epub", coverFileName: nil)
        store.updateProgress(relativePath: "voyage.epub", locatorJSON: #"{"href":"c1","type":"application/xhtml+xml"}"#, progression: 0.4)

        let stored = store.book(atRelativePath: "voyage.epub")
        XCTAssertEqual(stored?.progression, 0.4)
        XCTAssertEqual(stored?.locatorJSON?.contains("c1"), true)
    }

    func testUpsertKeepsExistingLocatorJSON() {
        let store = LocalBookService()
        let fileURL = URL(fileURLWithPath: "/tmp/Books/voyage.epub")
        let item = LibraryItem(fileURL: fileURL)

        store.upsert(item, relativePath: "voyage.epub", coverFileName: nil)
        store.updateProgress(
            relativePath: "voyage.epub",
            locatorJSON: #"{"href":"c1","type":"application/xhtml+xml"}"#,
            progression: 0.4
        )
        store.upsert(item, relativePath: "voyage.epub", coverFileName: nil)

        let stored = store.book(atRelativePath: "voyage.epub")
        XCTAssertEqual(stored?.locatorJSON?.contains("c1"), true)
    }
}

File: android/app/src/test/java/miko/biblossurfer/data/BookStoreTests.kt
package miko.biblossurfer.data

import miko.biblossurfer.data.model.LibraryItem
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.File

class BookStoreTests {
    @Test
    fun progressRoundTripsThroughTheStore() {
        val store = LocalBookService()
        val fileURL = File("/tmp/Books/voyage.epub")
        val item = LibraryItem(fileURL = fileURL)

        store.upsert(item, "voyage.epub", null)
        store.updateProgress(
            relativePath = "voyage.epub",
            locatorJSON = """{"href":"c1","type":"application/xhtml+xml"}""",
            progression = 0.4,
        )

        val stored = store.book("voyage.epub")
        assertEquals(0.4, stored?.progression)
        assertTrue(stored?.locatorJSON?.contains("c1") == true)
    }

    @Test
    fun upsertKeepsExistingLocatorJSON() {
        val store = LocalBookService()
        val fileURL = File("/tmp/Books/voyage.epub")
        val item = LibraryItem(fileURL = fileURL)

        store.upsert(item, "voyage.epub", null)
        store.updateProgress(
            relativePath = "voyage.epub",
            locatorJSON = """{"href":"c1","type":"application/xhtml+xml"}""",
            progression = 0.4,
        )
        store.upsert(item, "voyage.epub", null)

        val stored = store.book("voyage.epub")
        assertTrue(stored?.locatorJSON?.contains("c1") == true)
    }
}
